Lib/fontTools/colorLib/geometry.py

```diff
@@ -0,0 +1,139 @@
+        self.centre = centre
+        self.radius = radius
+
+    def __repr__(self):
+        return f"Circle(centre={self.centre}, radius={self.radius})"
+
+    def round(self):
+        return Circle(_round_point(self.centre), otRound(self.radius))
+
+    def inside(self, outer_circle):
+        dist = self.radius + hypot(*_vector_between(self.centre, outer_circle.centre))
+        return (
+            abs(outer_circle.radius - dist) <= _NEARLY_ZERO
+            or outer_circle.radius > dist
+        )
+
+    def concentric(self, other):
+        return self.centre == other.centre
+
+    def nudge_towards(self, direction):
+        self.centre = _nudge_point(self.centre, direction)
+
+
+def round_start_circle_stable_containment(c0, r0, c1, r1):
+    """Round start circle so that it stays inside/outside end circle after rounding.
+
+    The rounding of circle coordinates to integers may cause an abrupt change
+    if the start circle c0 is so close to the end circle c1's perimiter that
+    it ends up falling outside (or inside) as a result of the rounding.
+    To keep the gradient unchanged, we nudge it in the right direction.
+
+    See:
+    https://github.com/googlefonts/colr-gradients-spec/issues/204
+    https://github.com/googlefonts/picosvg/issues/158
+    """
+    start_circle, end_circle = Circle(c0, r0), Circle(c1, r1)
+
+    inside_before_round = start_circle.inside(end_circle)
+
+    round_start = start_circle.round()
+    round_end = end_circle.round()
+
+    # At most 3 iterations ought to be enough to converge. In the first, we
+    # check if the start circle keeps containment after normal rounding; then
+    # we continue adjusting by -/+ 1.0 until containment is restored.
+    # Normal rounding can at most move each coordinates -/+0.5; in the worst case
+    # both the start and end circle's centres and radii will be rounded in opposite
+    # directions, e.g. when they move along a 45 degree diagonal:
+    #   c0 = (1.5, 1.5) ===> (2.0, 2.0)
+    #   r0 = 0.5 ===> 1.0
+    #   c1 = (0.499, 0.499) ===> (0.0, 0.0)
+    #   r1 = 2.499 ===> 2.0
+    # In this example, the relative distance between the circles, calculated
+    # as r1 - (r0 + distance(c0, c1)) is initially 0.57437 (c0 is inside c1), and
+    # -1.82842 after rounding (c0 is now outside c1). Nudging c0 by -1.0 on both
+    # x and y axes moves it towards c1 by hypot(-1.0, -1.0) = 1.41421. Two of these
+    # moves cover twice that distance, which is enough to restore containment.
+    max_attempts = 3
+    for _ in range(max_attempts):
+        inside_after_round = round_start.inside(round_end)
+        if inside_before_round == inside_after_round:
+            break
+        if round_start.concentric(round_end):
+            # can't move c0 towards c1 (they are the same), so we change the radius
+            if inside_after_round:
+                round_start.radius += 1.0
+            else:
+                round_start.radius -= 1.0
```

Comment:
I don't think you should change direction in the loop; you never want to move one step forward and one step back. Compute step outside and in the loop move by step.

Applies to both radius adjustment and vector nudges.

Reply:
but I'm not changing direction

Reply:
protect against radius becoming <= 0?

Reply:
it's ok if it's 0. Less than 0 not ok, I'll protect against that

Reply:
If you make the step a function your loop is something like:

if inside_before_round == round_start.inside(round_end)
  break # solved!
step() # step doesn't change direction, it always bumps radius or nudges in a given direction that was determined pre-loop

Reply:
Re changing direction, your loop has the ability to add one to radius one spin and subtract one the next. I don't think that's ever desired; you should always make the same adjustment each spin.

Reply:
I can't determine pre-loop because I'm checking `inside_after_round = round_start.inside(round_end)` _inside_ the loop.
It's impossible to flip direction! The loop stops as soon as `inside_before_round == inside_after_round`; it continues only if they are `!=`. It can't nudge in one direction, still be `inside_before_round != inside_after_round` and subsequently nudge in the opposite direction.

Reply:
I changed it to determine the direction to nudge outside of the loop, and placed an assert against negative radius (which I can't even reproduce empirically). I hope that addesses all the remaining issues.